### PR TITLE
Add sizers for collection reusable views 📏

### PR DESCRIPTION
**Pull request by filipe-lemos**
### Checklist
- [x] I've rebased my changes on top of `master`
- [x] I've built and run the project to see all new and existing tests pass
- [x] I've followed the [Mindera swift style guide](https://github.com/Mindera/swift-style-guide)
- [x] I've read the [Contribution Guidelines](https://github.com/Mindera/Alicerce/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
When laying out collection view cells (and other reusable views) it might be necessary to know the size of a cell before it is created.

### Description
The sizer is used to calculate the size of a view before it is created and added to the view hierarchy. This is useful in the case of collection view cells when the layout needs to know the size for a cell given its view model. It works by wrapping `systemLayoutSizeFitting` on a dummy view to which the view model is temporarily bound.

This also includes a cache to avoid repeating the same calculations unnecessarily when a previous value is known to be the same. It uses a dictionary of size values with a key generated by the view (conforming to `SizerViewModelView`) for a specific view model.

### CODE REVIEW

File: Sources/View/CollectionReusableViewSizer.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    }
+}
+
+final class CollectionReusableViewSizerCacheGroup {
+
+    private var sizers = [String: AnyObject]()
+
+    func sizeFor<View: UICollectionReusableView & SizerViewModelView>(
+        _ type: View.Type,
+        viewModel: CollectionReusableViewSizer<View>.ViewModel,
```
Comment (p4checo) [inline Sources/View/CollectionReusableViewSizer.swift]: perhaps we can define a `typealias Sizer = CollectionReusableViewSizer<View>` like above to reduce some of this verbosity 💅

File: Sources/View/CollectionReusableViewSizer.swift [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    associatedtype SizerCacheKey: Hashable = ConstantSizerCacheKey
+
+    static func sizerCacheKeyFor(viewModel: ViewModel) -> SizerCacheKey
+}
+
+extension SizerViewModelView where SizerCacheKey == ConstantSizerCacheKey {
+
+    static func sizerCacheKeyFor(viewModel: ViewModel) -> SizerCacheKey {
+
+        return SizerCacheKey()
```
Comment (p4checo) [inline Sources/View/CollectionReusableViewSizer.swift]: maybe we can inline this 💄

File: Sources/View/CollectionReusableViewSizer.swift [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        width: Dimension = .compressed,
+        height: Dimension = .compressed,
+        attributes: UICollectionViewLayoutAttributes? = nil
+    ) -> CGSize {
+
+        assert(view.viewModel == nil, "🧟‍♂️ Zombie view model detected! Kill it with fire!")
+
+        view.viewModel = viewModel // the sizer populates the dummy view with the view model
+
+        attributes.then(f: { view.apply($0) })
```
Comment (p4checo) [inline Sources/View/CollectionReusableViewSizer.swift]: perhaps we can use a simple `attributes.flatMap(view.apply)`? 💅

**Comment by p4checo** [comment]
There are still some untested areas, it would be great if you could add some tests covering them 🙇

File: Sources/View/CollectionReusableViewSizer.swift (hunk 1 above)
Comment (filipe-lemos) [reply Sources/View/CollectionReusableViewSizer.swift]: Above the generic was on the type while here it's on the func so the typealias would also need to be generic:

```
    typealias Sizer<View: UICollectionReusableView & SizerViewModelView> = CollectionReusableViewSizer<View>

    func sizeFor<View: UICollectionReusableView & SizerViewModelView>(
        _ type: View.Type,
        viewModel: Sizer<View>.ViewModel,
        width: Sizer<View>.Dimension = .compressed,
        height: Sizer<View>.Dimension = .compressed,
        attributes: UICollectionViewLayoutAttributes? = nil
    ) -> CGSize {
```

Not sure we gain much with this, do you think it's still worth it?

Comment (p4checo) [reply Sources/View/CollectionReusableViewSizer.swift]: Oh, I see. Missed that 🙈 never mind then!

**Pull request merged**